x-pack/filebeat/input/azureeventhub/v2_input.go

```diff
@@ -0,0 +1,532 @@
+				in.log.Infow(
+					"ownership lost for partition, stopping processing",
+					"partition", partitionID,
+				)
+
+				return nil
+			}
+
+			return err
+		}
+
+		if len(events) == 0 {
+			continue
+		}
+
+		err = in.processReceivedEvents(events, partitionID, pipelineClient)
+		if err != nil {
+			return fmt.Errorf("error processing received events: %w", err)
+		}
+	}
+}
+
+// processReceivedEvents
+func (in *eventHubInputV2) processReceivedEvents(receivedEvents []*azeventhubs.ReceivedEventData, partitionID string, pipelineClient beat.Client) error {
+	processingStartTime := time.Now()
+	eventHubMetadata := mapstr.M{
+		"partition_id":   partitionID,
+		"eventhub":       in.config.EventHubName,
+		"consumer_group": in.config.ConsumerGroup,
+	}
+
+	for _, receivedEventData := range receivedEvents {
+		// Update input metrics.
+		in.metrics.receivedMessages.Inc()
+		in.metrics.receivedBytes.Add(uint64(len(receivedEventData.Body)))
+
+		// A single event can contain multiple records.
+		// We create a new event for each record.
+		records := in.messageDecoder.Decode(receivedEventData.Body)
+
+		for record := range records {
+			_, _ = eventHubMetadata.Put("offset", receivedEventData.Offset)
+			_, _ = eventHubMetadata.Put("sequence_number", receivedEventData.SequenceNumber)
+			_, _ = eventHubMetadata.Put("enqueued_time", receivedEventData.EnqueuedTime)
+
+			// The partition key is optional.
+			if receivedEventData.PartitionKey != nil {
+				_, _ = eventHubMetadata.Put("partition_key", *receivedEventData.PartitionKey)
+			}
+
+			event := beat.Event{
+				// this is the default value for the @timestamp field; usually the ingest
+				// pipeline replaces it with a value in the payload.
+				Timestamp: processingStartTime,
+				Fields: mapstr.M{
+					"message": record,
+					"azure":   eventHubMetadata,
+				},
+				Private: receivedEventData,
+			}
+
+			// Publish the event to the Beats pipeline.
+			pipelineClient.Publish(event)
+
+			// Update input metrics.
+			in.metrics.sentEvents.Inc()
+		}
+
+		// Update input metrics.
+		in.metrics.processedMessages.Inc()
+		in.metrics.processingTime.Update(time.Since(processingStartTime).Nanoseconds())
+	}
+
+	return nil
+}
+
+// initializePartitionResources initializes any partition specific resources for your application.
+//
+// Sets up a pipelineClient for publishing events and receive notification of their ACKs.
+func initializePartitionResources(ctx context.Context, partitionClient *azeventhubs.ProcessorPartitionClient, pipeline beat.Pipeline, log *logp.Logger) (beat.Client, error) {
+	// initialize things that might be partition specific, like a
+	// database connection.
+	return pipeline.ConnectWith(beat.ClientConfig{
+		EventListener: acker.LastEventPrivateReporter(func(acked int, data any) {
+			err := partitionClient.UpdateCheckpoint(ctx, data.(*azeventhubs.ReceivedEventData), nil)
+			if err != nil {
```

Comment:
Just curious, what is this checkpointing for?

If the checkpointing fails, no return? Is there a retry mechanism that we should introduce?

Example: https://sourcegraph.com/github.com/tsaikd/gogstash/-/blob/input/azureeventhub/inputazureeventhub.go?L165

Reply:
Should we try retries (w/ exponential backoff or simple equal duration spaced retries) to try checkpointing?

Reply:
> Just curious, what is this checkpointing for?

The checkpoint information stores the sequence number of the last message that the worker processed successfully for that partition. 

The input can continue from the last successful message in case of restart or partition ownership change.


> If the checkpointing fails, no return? Is there a retry mechanism that we should introduce?

> Should we try retries (w/ exponential backoff or simple equal duration spaced retries) to try checkpointing?

Good point. We can add some form of retry to make it more resilient.

However, if the checkpoint store continues to fail, we have a limited option here: this is the callback from the pipeline client calls when the output acks a delivery to ES.

Reply:
A couple of more review comments:

- The type assertion data.(*azeventhubs.ReceivedEventData) could panic if the type is incorrect. If we are 100% sure that is the type we are going to receive; let's capture the same in a comment?
- Do you think logging partitionClient's methods like PartitionID(), data.(*azeventhubs.ReceivedEventData).SequenceNumber, data.(*azeventhubs.ReceivedEventData).Offset because these are getting checkpointed. Beneficial during debugging?

Reply:
> The type assertion data.(*azeventhubs.ReceivedEventData) could panic if the type is incorrect. If we are 100% sure that is the type we are going to receive; let's capture the same in a comment?

Yeah, we are in control of the type we set:

```go
	event := beat.Event{
		// this is the default value for the @timestamp field; usually the ingest
		// pipeline replaces it with a value in the payload.
		Timestamp: processingStartTime,
		Fields: mapstr.M{
			"message": record,
			"azure":   eventHubMetadata,
		},
		Private: receivedEventData,
	}
```

However, it's better to check. I'm adding a check to make sure the type is correct. This will mainly help our future selves, making changes to this input down the road.

Reply:
> Do you think logging partitionClient's methods like PartitionID(), data.(*azeventhubs.ReceivedEventData).SequenceNumber, data.(*azeventhubs.ReceivedEventData).Offset because these are getting checkpointed. Beneficial during debugging?

Yeah, I think it's a good idea. 

Adding it.